### PR TITLE
Multiple file support

### PR DESCRIPTION
**Pull request by pbochis**

### CODE REVIEW

File: app/elements/coder/challenge/tasks/general-code-task.html [hunk 1]
```diff
@@ -1,11 +1,15 @@
 <link rel="import" href="/bower_components/iron-icons/av-icons.html">
 <link rel="import" href="/bower_components/paper-button/paper-button.html">
 <link rel="import" href="/bower_components/paper-material/paper-material.html">
+<link rel="import" href="/bower_components/iron-flex-layout/classes/iron-flex-layout.html">
+<link rel="import" href="/bower_components/iron-icons/av-icons.html">
+
 <link rel="import" href="/elements/coder/code-editor/code-editor.html">
 <link rel="import" href="/elements/coder/code-editor/language-selector.html">
+<link rel="import" href="/elements/coder/challenge/ws-element.html">
```
Comment (victorbalan) [inline app/elements/coder/challenge/tasks/general-code-task.html]: why?

File: app/elements/coder/code-editor/code-editor-with-actions.html [hunk 2]
```diff
@@ -65,13 +65,15 @@
 			}
 		</style>
 		<template-service id="templateService" template="{{template}}"></template-service>
+		<user-submission-service id="submissionService"></user-submission-service>
+		<ws-element url="{{wsUrl}}"></ws-element>
```
Comment (victorbalan) [inline app/elements/coder/code-editor/code-editor-with-actions.html]: why? ws is in base code editor

File: app/elements/coder/challenge/tasks/general-code-task.html [hunk 3]
```diff
@@ -77,7 +82,7 @@
 			<div class="flex">
 				<paper-material id="topRightCard" elevation="1">
 					<div id="codeEditorWrapper">
-						<code-editor id="codeEditor" task-id="{{task.id}}" language="{{language.value}}"></code-editor>
```
Comment (victorbalan) [inline app/elements/coder/challenge/tasks/general-code-task.html]: why remove task id?

File: app/elements/coder/challenge/tasks/general-code-task.html [hunk 4]
```diff
@@ -116,14 +121,18 @@
...
 					notify: true
+				},
+				language: {
+					type: Object,
+					notify: true
 				}
 			},
 			runCode: function(){
-				this.$.codeEditor.run();
+				this.$.submissionService.run(this.task.id, this.language.value, this.$.codeEditor.getFiles());
```
Comment (victorbalan) [inline app/elements/coder/challenge/tasks/general-code-task.html]: this is code editor's job. please put it back where it belongs

File: app/elements/coder/challenge/tasks/general-code-task.html [hunk 5]
```diff
@@ -116,14 +121,18 @@
...
 				this.control.loading();
 			},
 			runTests: function(){
-				this.$.codeEditor.submit();
+				this.$.submissionService.submit(this.task.id, this.language.value, this.$.codeEditor.getFiles());
```
Comment (victorbalan) [inline app/elements/coder/challenge/tasks/general-code-task.html]: this is code editor's job. please put it back where it belongs

File: app/elements/coder/code-editor/code-editor.html [hunk 6]
```diff
@@ -26,70 +49,109 @@
...
+				this._addContextMenuListener();
+				//initialize code editor
 				ace.require('ace/ext/language_tools');
 				this.$.editor = ace.edit(this.$.editor);
 				this.$.editor.setTheme('ace/theme/chrome');
 				this.$.editor.setOptions({
 					enableBasicAutocompletion: true
 				});
 				var self = this;
-				this.$.editor.commands.addCommand({
```
Comment (victorbalan) [inline app/elements/coder/code-editor/code-editor.html:60]: why did you remove this?

File: app/elements/coder/code-editor/code-editor.html [hunk 7]
```diff
@@ -26,70 +49,109 @@
...
 				},
-				taskId: {
+				toggleExplorerIcon: {
 					type: String,
-					reflectToAttribute: true
+					notify: true,
+					value: 'chevron-left'
 				},
-				language: {
+				mimeType: String,
```
Comment (victorbalan) [inline app/elements/coder/code-editor/code-editor.html]: you don`t use this anymore

File: app/elements/coder/code-editor/file-explorer/context-menu.html [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+			},
+			showMenu: function(data){
+				if (this.pos.x === data.x && this.pos.y === data.y){
+					return;
+				}
+				this.contextStyle = '';
+				this.selectedItem = data.element;
+				this.notifyPath('selectedItem.meta.isFolder', this.selectedItem.meta.isFolder);
+				this.isRoot = this.selectedItem.meta.isFolder && this.selectedItem.meta.name === 'root';
+				console.log(this.isRoot);
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/context-menu.html]: d

File: app/elements/coder/code-editor/file-explorer/context-menu.html [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+				this._updatePosition(data);
+			},
+			_updatePosition: function(data){
+				if (!this.pos){
+					this.pos = {};
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/context-menu.html]: if you init it with `{}` and not return why do you need the if?

File: app/elements/coder/code-editor/code-editor.html [hunk 10]
```diff
@@ -26,70 +49,109 @@
...
+					this.selectedFile.markUnselected();
+					this.selectedFile = null;
+				}
+			},
+			_fileSelected: function(e, detail){
+				if (!!this.selectedFile){
+					this.selectedFile.markUnselected();
+				}
+				this.selectedFile = detail;
+				this.selectedFile.markSelected();
```
Comment (victorbalan) [inline app/elements/coder/code-editor/code-editor.html]: you can do this where you fire the event

File: app/elements/coder/code-editor/file-explorer/explorer-file.html [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+			listeners: {
+				'click': 'selectFile'
+			},
+			_afterMetaSet: function(){
+				if (this.meta.selected){
+					this.selectFile();
+				}
+			},
+			selectFile: function(e){
+				this.fire('file-selected', this);
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-file.html]: you can add here `this.markSelected()`

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html [hunk 12]
```diff
@@ -0,0 +1,182 @@
+<link rel="import" href="/bower_components/iron-flex-layout/iron-flex-layout-classes.html">
+<link rel="import" href="/bower_components/iron-icon/iron-icon.html">
+<link rel="import" href="/bower_components/paper-toast/paper-toast.html">
+
+<link rel="import" href="explorer-file.html">
+
+<dom-module id="explorer-folder">
+	<template>
+		<style include="iron-flex">
+			.folder-wrapper:hover {
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: you can merge this style + file style into an explorer-item style

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+		<paper-toast id="toast"></paper-toast>
+	</template>
+	<script type="application/javascript" src="/scripts/behaviors/file-explorer-item-behavior.js"></script>
+	<script>
+		Polymer({
+			is: 'explorer-folder',
+			behaviors: [
+				Behaviors.FileExplorerItemBehavior
+			],
+			addFolder: function(folderName){
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: you can add `editMode = true` by default so the user does not rightclick and rename.

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+			findFirstAvailableName(name, isFolder){
+				var ok = false;
+				var tries = 0;
+				var searchName = name;
+				while(!ok){
+					ok = true;
+					for (var i = 0; i < this.meta.children.length; i++){
+						if (this.meta.children[i].name === searchName && this.meta.children[i].isFolder === isFolder){
+							tries++;
+							searchName = name + '(' + tries + ')';
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: why don`t you construct this outside the for and while?

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+			},
+			_checkRename: function(e, el){
+				var conflict = this.findFirstAvailableName(el.newName, el.meta.isFolder) !== el.newName;
+				el.finishRename(conflict);
+			},
+			findFirstAvailableName(name, isFolder){
+				var ok = false;
+				var tries = 0;
+				var searchName = name;
+				while(!ok){
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: can you explain the point of this while? You iterate the exact same collection everytime

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html (hunk 14 above)
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: also..this would not be needed if you do editmode true by default. You would just need an existing check.

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+				}
+				var files = [];
+				for (var i = 0; i < this.meta.children.length; i++){
+					var el = this.meta.children[i];
+					if (!el.isFolder){
+						files.push({
+							'name': el.name,
+							'content': el.content
+						});
+						console.log(files);
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: d

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+			_deleteChild: function(e, childMeta){
+				var index = this.meta.children.indexOf(childMeta);
+				if (index < 0){
+					return;
+				}
+				this.splice('meta.children', index, 1);
+			},
+			selectFile: function(){
+				var file = this.$$('explorer-file');
+				console.log(file);
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: d

File: app/elements/coder/code-editor/file-explorer/file-explorer.html [hunk 18]
```diff
@@ -0,0 +1,78 @@
+<link rel="import" href="/bower_components/iron-flex-layout/iron-flex-layout-classes.html">
+<link rel="import" href="/bower_components/paper-material/paper-material.html">
+
+<link rel="import" href="explorer-folder.html">
+<link rel="import" href="context-menu.html">
+
+<dom-module id="file-explorer">
```
Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/file-explorer.html]: where is this used?

File: app/scripts/util.js [hunk 19]
```diff
@@ -85,10 +85,18 @@ var util = {
 
 		return result + ' ' + util.padTime(date.getUTCHours()) + ':' + util.padTime(date.getUTCMinutes()) + ' UTC';
 	},
-	languages: [
-		'java',
-		'py',
-		'c',
-		'js'
-	]
+	clickInsideElement: function(e, elName){
```
Comment (victorbalan) [inline app/scripts/util.js]: why did u add this here if you use it just in on file. If you want it to be reusable pls do a behavior. Soon we'll ditch the `util.js` file.

File: app/elements/coder/code-editor/file-explorer/explorer-folder.html (hunk 12 above)
Comment (pbochis) [inline app/elements/coder/code-editor/file-explorer/explorer-folder.html]: awesome sugestion! thanks

**Comment by victorbalan** [comment]
will be pushed by @flowlo
